src/plugins/workspace/public/components/workspace_form/utils.ts

```diff
@@ -98,3 +115,219 @@ export const convertApplicationsToFeaturesOrGroups = (
     ];
   }, []);
 };
+
+export const isUserOrGroupPermissionSettingDuplicated = (
+  permissionSettings: Array<Partial<WorkspacePermissionSetting>>,
+  permissionSettingToCheck: WorkspacePermissionSetting
+) =>
+  permissionSettings.some(
+    (permissionSetting) =>
+      (permissionSettingToCheck.type === WorkspacePermissionItemType.User &&
+        permissionSetting.type === WorkspacePermissionItemType.User &&
+        permissionSettingToCheck.userId === permissionSetting.userId) ||
+      (permissionSettingToCheck.type === WorkspacePermissionItemType.Group &&
+        permissionSetting.type === WorkspacePermissionItemType.Group &&
+        permissionSettingToCheck.group === permissionSetting.group)
+  );
+
+// default permission mode is read
+export const getPermissionModeId = (modes: WorkspacePermissionMode[]) => {
+  for (const key in optionIdToWorkspacePermissionModesMap) {
+    if (optionIdToWorkspacePermissionModesMap[key].every((mode) => modes?.includes(mode))) {
+      return key;
+    }
+  }
+  return PermissionModeId.Read;
+};
+
+export const convertPermissionSettingsToPermissions = (
+  permissionItems: WorkspacePermissionSetting[] | undefined
+) => {
+  if (!permissionItems || permissionItems.length === 0) {
+    return undefined;
+  }
+  return permissionItems.reduce<SavedObjectPermissions>((previous, current) => {
+    current.modes.forEach((mode) => {
+      if (!previous[mode]) {
+        previous[mode] = {};
+      }
+      switch (current.type) {
+        case 'user':
```

Comment:
Should we add more error handling here?

````
              if (current.userId) {
              ....
              } else {
                throw new Error('Missing userId in permission setting');
              }
````

Reply:
The `convertPermissionSettingsToPermissions` is only be used in workspace form submit handler. The passed permissionItems will pass the form data validation and will have `userId` property.  Since we already define the permissionItems as `WorkspacePermissionSetting[]` type. The userId is required when type equal 'user', I think we don't need to add this error handling.